### PR TITLE
Add torch DiscreteCNNModule & DiscreteCNNQFunction

### PR DESCRIPTION
**Pull request by maliesa96**
Adds the mentioned primitives to torch. This also prevents `CNNModule` from mutating the input tensor when normalizing images, which it was previously doing.

### CODE REVIEW

**Review by ryanjulian — APPROVED** [review]
Please fixup the codecov issue

File: tests/garage/torch/modules/test_discrete_cnn_module.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+import numpy as np
+import pytest
+import torch
+import torch.nn as nn
+
+from garage.torch.modules import CNNModule, DiscreteCNNModule, MLPModule
+
+
+class TestDiscreteCNNModule:
```
Comment (ryanjulian) [inline tests/garage/torch/modules/test_discrete_cnn_module.py:12]: can you use a fixture and a pytest style test? we have been trying to discourage new xunit-style tests

File: tests/garage/torch/q_functions/test_discrete_cnn_q_function.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import torch
+from torch import nn
+
+from garage.envs import GymEnv
+from garage.torch.q_functions import DiscreteCNNQFunction
+
+from tests.fixtures.envs.dummy import DummyBoxEnv
+
+
+class TestDiscreteCNNQFunction:
```
Comment (ryanjulian) [inline tests/garage/torch/q_functions/test_discrete_cnn_q_function.py:13]: do these need xunit style? we've been trying to encourage new tests use pytest style.

File: src/garage/torch/modules/discrete_cnn_module.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+from garage.torch.modules import CNNModule, MLPModule
+
+
+# pytorch v1.6 issue, see https://github.com/pytorch/pytorch/issues/42305
+# pylint: disable=abstract-method
+class DiscreteCNNModule(nn.Module):
+    """Discrete CNN Q Function.
+
+    A Q network that estimates Q values of discrete actions. It is constructed
```
Comment (yeukfu) [inline src/garage/torch/modules/discrete_cnn_module.py:13]: Docstring mismatch

File: src/garage/torch/modules/discrete_cnn_module.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    using a CNN followed by one or more fully-connected layers.
+
+    Args:
+        input_shape (tuple): Shape of the input. Based on 'NCHW' data format:
```
Comment (yeukfu) [inline src/garage/torch/modules/discrete_cnn_module.py:17]:
```suggestion
        input_shape (tuple[int]): Shape of the input. Based on 'NCHW' data format:
```

File: src/garage/torch/modules/discrete_cnn_module.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+        Args:
+            inputs (torch.Tensor): Inputs to the model of shape
+                (input_shape*).
+
+        Returns:
+            torch.Tensor: Output tensor of shape :math:`(N, output_dim)`.
+
+        """
+        return self._module(inputs)
```
Comment (yeukfu) [inline src/garage/torch/modules/discrete_cnn_module.py:136]: Need an empty line in the end?

File: src/garage/torch/q_functions/discrete_cnn_q_function.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                         paddings=paddings,
+                         padding_mode=padding_mode,
+                         max_pool=max_pool,
+                         pool_shape=pool_shape,
+                         pool_stride=pool_stride,
+                         output_nonlinearity=output_nonlinearity,
+                         output_w_init=output_w_init,
+                         output_b_init=output_b_init,
+                         layer_normalization=layer_normalization,
+                         is_image=is_image)
```
Comment (yeukfu) [inline src/garage/torch/q_functions/discrete_cnn_q_function.py:112]: Need an empty line in the end?

**Review by yeukfu — APPROVED** [review]

File: tests/garage/torch/modules/test_discrete_cnn_module.py (hunk 1 above)
Comment (maliesa96) [reply tests/garage/torch/modules/test_discrete_cnn_module.py:12]: Do you have an example? Is the difference simply that I shouldn't use a class?
